<a href="https://colab.research.google.com/github/Thakurvighnesh/HamoMLDatascience/blob/main/Hamoye_ML_StageB_Quiz.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Energy Data

In [1]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.preprocessing import MinMaxScaler
from sklearn.linear_model import Ridge
from sklearn.linear_model import Lasso

In [2]:
df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/00374/energydata_complete.csv')

In [3]:
df.head()

date  Appliances  lights  ...  Tdewpoint        rv1        rv2
0  2016-01-11 17:00:00          60      30  ...        5.3  13.275433  13.275433
1  2016-01-11 17:10:00          60      30  ...        5.2  18.606195  18.606195
2  2016-01-11 17:20:00          50      30  ...        5.1  28.642668  28.642668
3  2016-01-11 17:30:00          50      40  ...        5.0  45.410389  45.410389
4  2016-01-11 17:40:00          60      40  ...        4.9  10.084097  10.084097

[5 rows x 29 columns]

In [4]:
scaler = MinMaxScaler()

drop_df = df.drop(columns=['date', 'lights'])
normalised_df = pd.DataFrame(scaler.fit_transform(drop_df), columns=drop_df.columns)
features_df = normalised_df.drop(columns=['Appliances'])
targets_df = normalised_df['Appliances']

x_train, x_test, y_train, y_test = train_test_split(features_df['T2'].values.reshape(-1,1), features_df['T6'], test_size=0.3, random_state=42)

linear_model = LinearRegression()

In [ ]:
features_df.head()

T1      RH_1        T2  ...  Tdewpoint       rv1       rv2
0  0.32735  0.566187  0.225345  ...   0.538462  0.265449  0.265449
1  0.32735  0.541326  0.225345  ...   0.533937  0.372083  0.372083
2  0.32735  0.530502  0.225345  ...   0.529412  0.572848  0.572848
3  0.32735  0.524080  0.225345  ...   0.524887  0.908261  0.908261
4  0.32735  0.531419  0.225345  ...   0.520362  0.201611  0.201611

[5 rows x 26 columns]

In [5]:
linear_model.fit(x_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [10]:
predictions = linear_model.predict(x_test)
rsquared = metrics.r2_score(y_test, predictions)
print("R^2", round(rsquared, 2))
#Q12:

R^2 0.64


In [11]:
mae = metrics.mean_absolute_error(y_test, predictions)
print("Mean Absolute Error: ", round(mae, 2))
#Q13:

Mean Absolute Error:  0.08


In [12]:


rss = np.sum(np.square(y_test - predictions))
print("Residual Sum of Squares: ", round(rss, 2))
#Q14:

Residual Sum of Squares:  66.12


In [13]:
rmse = np.sqrt(metrics.mean_squared_error(y_test, predictions))
print("RMSE: ", round(rmse, 3))
#Q15:

RMSE:  0.106


In [14]:


coe_det = metrics.r2_score(y_test, predictions)
print("Coef of D: ", round(coe_det, 2))

#Q16:

Coef of D:  0.64


In [15]:


X = normalised_df.drop(columns=['Appliances'])
Y = normalised_df['Appliances']

x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size=0.3, random_state=42)

linear_model_2 = LinearRegression()
linear_model_2.fit(x_train, y_train)

predictions_2 = linear_model_2.predict(x_test)

In [16]:
def get_weights_df(model, feat, col_name):
    weights = pd.Series(model.coef_, feat.columns).sort_values()
    weights_df = pd.DataFrame(weights).reset_index()
    weights_df.columns = ['Features', col_name]
    weights_df[col_name].round(3)
    return weights_df

In [17]:
linear_model_weights = get_weights_df(linear_model_2, x_train, 'Linear_Model_Weight')

In [18]:
linear_model_weights.sort_values(by='Linear_Model_Weight', ascending=True)
#Q17:RH_2,RH_1

Features  Linear_Model_Weight
0          RH_2            -0.456698
1         T_out            -0.321860
2            T2            -0.236178
3            T9            -0.189941
4          RH_8            -0.157595
5        RH_out            -0.077671
6          RH_7            -0.044614
7          RH_9            -0.039800
8            T5            -0.015657
9            T1            -0.003281
10          rv1             0.000770
11          rv2             0.000770
12  Press_mm_hg             0.006839
13           T7             0.010319
14   Visibility             0.012307
15         RH_5             0.016006
16         RH_4             0.026386
17           T4             0.028981
18    Windspeed             0.029183
19         RH_6             0.038049
20         RH_3             0.096048
21           T8             0.101995
22    Tdewpoint             0.117758
23           T6             0.236425
24           T3             0.290627
25         RH_1             0.553547

In [19]:
lasso_model = Lasso(alpha=0.001)
lasso_model.fit(x_train, y_train)

Lasso(alpha=0.001, copy_X=True, fit_intercept=True, max_iter=1000,
      normalize=False, positive=False, precompute=False, random_state=None,
      selection='cyclic', tol=0.0001, warm_start=False)

In [22]:
lasso_weights = get_weights_df(lasso_model, x_train, 'Lasso_Model_Weight')
lasso_weights[lasso_weights['Lasso_Model_Weight'] != 0]
#Q19:

Features  Lasso_Model_Weight
0      RH_out           -0.049557
1        RH_8           -0.000110
24  Windspeed            0.002912
25       RH_1            0.017880

The given information appears to be a table representing features and their corresponding weights obtained from a Lasso regression model. The Lasso regression model is a type of linear regression that incorporates regularization by adding a penalty term based on the absolute values of the coefficients. This penalty encourages sparsity in the model by driving some coefficients to zero, effectively performing feature selection.

Let's analyze the features and their Lasso model weights:

RH_out: The feature "RH_out" has a weight of -0.049557. A negative weight indicates that as the value of this feature increases, the predicted outcome (dependent variable) tends to decrease. In this case, the model suggests that an increase in the outside humidity (RH_out) is associated with a decrease in the predicted outcome.

RH_8: The feature "RH_8" has a weight of -0.000110. This weight is close to zero, suggesting that the model found little or no significant relationship between this feature and the predicted outcome. In other words, the Lasso model determined that "RH_8" does not contribute much to the prediction.

Windspeed: The feature "Windspeed" has a weight of 0.002912. A positive weight indicates that as the value of this feature increases, the predicted outcome tends to increase as well. The model suggests that higher windspeed is associated with a higher predicted outcome.

RH_1: The feature "RH_1" has a weight of 0.017880. Similar to the "Windspeed" feature, a positive weight implies that an increase in "RH_1" corresponds to an increase in the predicted outcome.

Conclusion:
Based on the Lasso model weights, we can infer the following:

The outside humidity (RH_out) and the initial humidity (RH_1) have positive relationships with the predicted outcome. Higher values of these features tend to be associated with higher predicted outcomes.
Windspeed also has a positive relationship with the predicted outcome, suggesting that higher windspeed is associated with higher predicted outcomes.
The feature RH_8 has a weight close to zero, indicating that it is not strongly associated with the predicted outcome according to the Lasso model.
It is important to note that the interpretation of these weights should be considered within the context of the specific dataset and the assumptions of the Lasso regression model. Additionally, further analysis and evaluation, including considering other regression metrics and potentially cross-validation, would be necessary to fully assess the performance and significance of the model and its features.